### PR TITLE
jupyter notebook widgets and interactive plotting

### PR DESCRIPTION
**Pull request by Chinmay4400**
This PR is in relation to the project jupyter notebook widget. I have plotted the spectrum graph interactively using bqplot and ipywidgets.

## Description

Using bqplot and ipywidgets, I plotted the spectrum graph as seen in quickstart notebook, interactively. In a separate notebook, I ran the simulation and plotted the graph such that user is allowed to print any of the three plots in any combination.
I added check box widgets to allow selection of plots and a 'show' button to render the changes.
I have added documentation for the same.

## Motivation and Context

This PR is a step in the direction of migrating the qt GUI to jupyter notebook where it is better and easily accessible by users.

## Screencast :
Peek 2020-03-16 23-04

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by Chinmay4400** [comment]
Enabled panning and zooming in bqplot but this is a bit unresponsive. Instead I'm gonna try to achieve this via ipywidgets

**Comment by Chinmay4400** [comment]
@jaladh-singhal Please review

**Comment by jaladh-singhal** [comment]
@chinmay0301 I can see your code but not the plots, unless I run it locally (which will require some time). Could you please attach the screen-casts (GIFs) of your interactive plots - make sure to interact with them when screen-casting?

File: docs/interactive/tardis_example.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# Example YAML configuration for TARDIS
```
Comment (jaladh-singhal) [inline docs/interactive/tardis_example.yml]: What's the purpose of this file as per you?

Comment (Chinmay4400) [reply docs/interactive/tardis_example.yml]: I figured the user won't have to download the yml file separately. Should I remove it?

Comment (jaladh-singhal) [reply docs/interactive/tardis_example.yml]: Yes you can if they don't concern with the work you did in notebook.

**Comment by jaladh-singhal** [comment]
That's great! What's `xy..` button for - panning along both x & y?

Also, I see you have not implemented zooming - you can try that as next step!

**Comment by Chinmay4400** [comment]
I was trying to get panning in x-axis, y-axis and xy both, other than the toolbar that bqplot provides.
I have implemented zooming but the interface is terrible. I am looking for other ways to enable zooming. Probably slider based or button based

**Comment by Chinmay4400** [comment]
Also, for the past two days I have been working with plotly on a side project. If we're not bounded to bqplot, I'd like to give it a try

**Comment by jaladh-singhal** [comment]
Sure! Even in 1st objective, I mentioned bokeh because it's much more mature than bqplot and same goes for plotly. Plotly is indeed great choice, go with it - but since it will make things easier, I would expect you to add some more functionalities other than this simple quickstart plot. :wink: 

Take inspiration from GUI guide in TARDIS docs, and show us what you can do (Aim is to re-build GUI using Jupyter ecosystem).

**Comment by Chinmay4400** [comment]
With plotly I'll try to get as much of GUI as possible. Also, did you get a chance to look at plasma graph?

**Comment by jaladh-singhal** [comment]
Sorry, no - I'll do it by tomorrow for sure!

**Comment by Chinmay4400** [comment]
@jaladh-singhal I have a doubts regarding the shell plot here. I'm pretty sure the axes of the ellipses are not dependent on any parameter and only the colors are representing any info. I'm working on a way to create this ellipse, just wanted to know if I'm thinking right.
image

**Comment by Chinmay4400** [comment]
New Graphs added, Yet to add widgets to toggle between graphs and interconnect these graphs
Peek 2020-03-16 23-04

**Comment by Chinmay4400** [comment]
I have an idea to draw the shell spectra but I have a doubt regarding their shape. Is the shape related to some data they represent, or is it arbitrary.
I have checked out the parameters radius,velocity etc. but they all suggest that the shape should be concentric circles. Can you point me in the right direction?

**Comment by Chinmay4400** [comment]
Also, looking over the TEP for  this  project, I think we should consider using bokeh server to interact with our model as jupyter notebook has serious space constraints. I'm still reading up on bokeh server so I 'm not sure if this is relevant to us. Please let me know what you think. 
https://towardsdatascience.com/data-visualization-with-bokeh-in-python-part-iii-a-complete-dashboard-dc6a86aa6e23

**Comment by jaladh-singhal** [comment]
> I think we should consider using bokeh server to interact with our model as jupyter notebook has serious space constraints.

@Chinmay4400 Could be more clear what do you mean?

BTW, if you are using plotly, you should check `figurewidgets` - they are what you'll most probably need to interface Jupyter widgets with interactive features of plotly plots. And if you want to use bokeh, then bokeh server is good to go (but you haven't shown us anything using bokeh).

**Comment by ycamacho** [comment]
> @jaladh-singhal I have an idea to draw the shell spectra but I have a doubt regarding their shape. Is the shape related to some data they represent, or is it arbitrary.
> I have checked out the parameters radius,velocity etc. but they all suggest that the shape should be concentric circles. Can you point me in the right direction?

Hi @Chinmay4400, by using concentric circles we are assuming that the explosion is spherically symmetric. One way to think about it is that some stars are like onions with layers of different elements and when the star explodes the ejecta has a similar spherical structure with different layers. This is a really simplified picture but it's a good assumption TARDIS makes in order to be able to model the spectra within a short time.  If you want to learn more about this, I'll point you to this article https://www.nature.com/articles/d41586-018-04601-7#

**Comment by Chinmay4400** [comment]
> > I think we should consider using bokeh server to interact with our model as jupyter notebook has serious space constraints.
> 
> @Chinmay4400 Could be more clear what do you mean?
> 
> BTW, if you are using plotly, you should check `figurewidgets` - they are what you'll most probably need to interface Jupyter widgets with interactive features of plotly plots. And if you want to use bokeh, then bokeh server is good to go (but you haven't shown us anything using bokeh).

I mean that if we make a full dashboard like the one in qt-GUI jupyter notebooks don't have enough space to show it properly. Bokeh server (or something similar; plotly uses dash) will make use of a browser window and give a proper dashboard.

**Comment by Chinmay4400** [comment]
Bokeh server is just an idea to get a dashboard in browser window, we can also use dash with plotly and see which works better.

**Comment by Chinmay4400** [comment]
@ycamacho Great Thanks! I was confused because the shape in QT-GUI appeared elliptical.

**Comment by Chinmay4400** [comment]
@jaladh-singhal @ycamacho Can I get a little bit more elaboration on these lines (from TEP) :

> Eventually, the GUI should also be able to not only work interactively, but also with stored models/simulation objects. This will require a helper which builds the simulation object, used by the GUI, from the stored information.

**Comment by Chinmay4400** [comment]
> Eventually, the GUI should also be able to not only work interactively, but also with stored models/simulation objects. This will require a helper which builds the simulation object, used by the GUI, from the stored information.
@wkerzendorf @ycamacho Can I get a little more clarification on this point from TEP?

**Comment by Chinmay4400** [comment]
This is a sample of what the dashboard will ultimately function like : 
Peek 2020-03-16 23-04
I have yet to figure out some caveats of nesting subplots in plotly

**Pull request closed without merging**